### ISSUE TITLE
Need a way to skip examples entirely

### ISSUE THREAD

**Issue by ciaranmcnulty** [phpspec/phpspec#605]
I found myself needing to skip an example entirely, because the collaborators are not available on different environments: https://github.com/ciaranmcnulty/phpspec/blob/refactor/support-questionhelper/src/PhpSpec/Console/Prompter/Question.php

The skipping was simple thanks to @gquemener's work in this area, but the problem is that it's then not possible to type hint the collaborator in the normal way.

What would be great would be a way to indicate that an entire example can be completely skipped at the Locator level. Whether that's via Annotations, config, some sort of programmatic way (like `MatcherProviderInterface`)

**Comment by gquemener** [comment]
Hello,

Not sure this can help, but I wrote an extension to skip examples through annotation: https://github.com/akeneo/PhpSpecSkipExampleExtension

**Comment by ciaranmcnulty** [comment]
Oh, very nice! I wonder if people would use it responsibly enough that it could be in core.

**Comment by ciaranmcnulty** [comment]
I rubber-ducked this with @MarcelloDuarte this morning and talked myself out of it.

It seems to me the use case for skipping examples depending on object availability is for writing integration tests. The example in the linked PR is an integration test written in PhpSpec, which is why it's painful.

I'm for keeping this pain for people attempting to use the tool for integration testing. I will change my PR to use some other method (possibly PHPUnit) to integration test those components that are strongly coupled to Symfony components.

I appreciate that some people are using the tool for functional or integration testing, so this extension will be useful for them; @gquemener can you please add a link to your extension in the docs at https://github.com/phpspec/phpspec/blob/master/docs/cookbook/extensions.rst ?

**Comment by stof** [comment]
which linked PR are you talking about ?

**Comment by stof** [comment]
> I appreciate that some people are using the tool for functional or integration testing, so this extension will be useful for them; @gquemener can you please add a link to your extension in the docs at https://github.com/phpspec/phpspec/blob/master/docs/cookbook/extensions.rst ?

I agree about linking to the extension, but I think it deserves to put a warning near it to educate people that they should use a different tool in this case

**Comment by ciaranmcnulty** [comment]
@stof the code I linked to in my first comment is #604

**Comment by ciaranmcnulty** [comment]
@stof There are a few extensions on that page that are clearly for functional/integration tests - maybe they should indeed be in a separate section with a disclaimer?

**Comment by stof** [comment]
> maybe they should indeed be in a separate section with a disclaimer?

agreed.

**Comment by stof** [comment]
I still see 1 case where skipping might be legitimate: based on the PHP version, when you have some features available only for newer PHP version. I don't think we can consider the case of PHP as an integration test.
So maybe skipping based on the PHP version should be easier in code (keeping the case of collaborators for an extension). What do you think ?

**Comment by ciaranmcnulty** [comment]
That's a good point, there may also be things like HHVM detection.

The use case we most likely want to support is 'my code has a conditional and does different things on different versions of PHP'.

The case I wouldn't want to support is 'my code only works on PHP >5.4 but I want to run the suite on 5.3 and skip the things that won't work'.

I'm not sure how we can distinguish the two.

**Comment by gquemener** [comment]
@ciaranmcnulty done: https://github.com/phpspec/phpspec/pull/606

**Comment by stof** [comment]
> The case I wouldn't want to support is 'my code only works on PHP >5.4 but I want to run the suite on 5.3 and skip the things that won't work'.

This use case may be needed too. Imagine your project provides optional traits to help using it (but your code does not make them mandatory). Specs ensuring your traits are behaving well (through specing a fixture class using them) would not be able to run at all on 5.3. But all other parts of your lib may still be compatible with PHP 5.3 and you would like to be able to run your specs on 5.3 to ensure it.
Another case for that I see is for Prophecy to ensure that it is able to double a class using variadic functions of PHP 5.6+ (disclaimer: it does not yet: https://github.com/phpspec/prophecy/pull/125). We don't want to forbid running the specs of the library on older versions of PHP (this would even be a major issue), but the particular case of variadic functions does not make sense on PHP 5.5 and lower (it cannot exist) and it cannot run (PHP parse error)

**Comment by ciaranmcnulty** [comment]
Right, good points. I guess I can't see a downside to skipping based on version then.

**Comment by ciaranmcnulty** [comment]
An alternative to this would be Behat-style profiles

**Comment by sagikazarmark** [comment]
I like the way how PHPUnit solved: `@requires resourceType resource`

In PHPSpec it is probably not necessary to skip a test based on an OS (but who knows), but this allows to be flexible about skipping reasons.

**Comment by alexislefebvre** [comment]
It looks like it's a duplicate of #119
